services/src/test/java/io/grpc/services/HealthCheckingLoadBalancerFactoryTest.java

```diff
@@ -0,0 +1,1102 @@
+  @Mock
+  private BackoffPolicy backoffPolicy2;
+
+  private HealthCheckingLoadBalancerFactory hcLbFactory;
+  private LoadBalancer hcLbEventDelivery;
+
+  @Before
+  @SuppressWarnings("unchecked")
+  public void setup() throws Exception {
+    MockitoAnnotations.initMocks(this);
+
+    for (int i = 0; i < NUM_SUBCHANNELS; i++) {
+      HealthImpl healthImpl = new HealthImpl();
+      healthImpls[i] = healthImpl;
+      Server server =
+          InProcessServerBuilder.forName("health-check-test-" + i)
+          .addService(healthImpl).directExecutor().build().start();
+      servers[i] = server;
+      ManagedChannel channel =
+          InProcessChannelBuilder.forName("health-check-test-" + i).directExecutor().build();
+      channels[i] = channel;
+
+      EquivalentAddressGroup eag =
+          new EquivalentAddressGroup(new FakeSocketAddress("address-" + i));
+      eags[i] = eag;
+      List<EquivalentAddressGroup> eagList = Arrays.asList(eag);
+      eagLists[i] = eagList;
+    }
+    resolvedAddressList = Arrays.asList(eags);
+    
+    when(backoffPolicyProvider.get()).thenReturn(backoffPolicy1, backoffPolicy2);
+    when(backoffPolicy1.nextBackoffNanos()).thenReturn(11L, 21L, 31L);
+    when(backoffPolicy2.nextBackoffNanos()).thenReturn(12L, 22L, 32L);
+
+    hcLbFactory = new HealthCheckingLoadBalancerFactory(
+        origLbFactory, backoffPolicyProvider, clock.getTimeProvider());
+    final LoadBalancer hcLb = hcLbFactory.newLoadBalancer(origHelper);
+    // Make sure all calls into the hcLb is from the syncContext
+    hcLbEventDelivery = new LoadBalancer() {
+        @Override
+        public void handleResolvedAddressGroups(
+            final List<EquivalentAddressGroup> servers, final Attributes attributes) {
+          syncContext.execute(new Runnable() {
+              @Override
+              public void run() {
+                hcLb.handleResolvedAddressGroups(servers, attributes);
+              }
+            });
+        }
+
+        @Override
+        public void handleSubchannelState(
+            final Subchannel subchannel, final ConnectivityStateInfo stateInfo) {
+          syncContext.execute(new Runnable() {
+              @Override
+              public void run() {
+                hcLb.handleSubchannelState(subchannel, stateInfo);
+              }
+            });
+        }
+
+        @Override
+        public void handleNameResolutionError(Status error) {
+          throw new AssertionError("Not supposed to be called");
+        }
+
+        @Override
+        public void shutdown() {
+          throw new AssertionError("Not supposed to be called");
+        }
+      };
+    verify(origLbFactory).newLoadBalancer(any(Helper.class));
+  }
+
+  @After
+  public void teardown() throws Exception {
+    // All scheduled tasks have been accounted for
+    assertThat(clock.getPendingTasks()).isEmpty();
+    // Health-check streams are usually not closed in the tests.  Force closing for clean up.
```

Comment:
Maybe mention that the cause is because handleSubchannelState is faked.

Reply:
Done.